Sources/OktaIdxAuth/Internal/Implementation.swift

```diff
@@ -0,0 +1,259 @@
+/*
+ * Copyright (c) 2021, Okta, Inc. and/or its affiliates. All rights reserved.
+ * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
+ *
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *
+ * See the License for the specific language governing permissions and limitations under the License.
+ */
+
+import Foundation
+import OktaIdx
+
+protocol OktaIdxAuthImplementation {
+    var delegate: OktaIdxAuthImplementationDelegate? { get set }
+    
+    func authenticate(username: String,
+                      password: String?,
+                      completion: OktaIdxAuth.ResponseResult<OktaIdxAuth.Response>?)
+    
+    func changePassword(_ password: String,
+                        completion: OktaIdxAuth.ResponseResult<OktaIdxAuth.Response>?)
+    
+    func recoverPassword(username: String,
+                         authenticator type: OktaIdxAuth.AuthenticatorType,
+                         completion: OktaIdxAuth.ResponseResult<OktaIdxAuth.Response>?)
+    
+    func verifyAuthenticator(code: String,
+                             completion: OktaIdxAuth.ResponseResult<OktaIdxAuth.Response>?)
+    
+    func revokeTokens(token: String,
+                      type: OktaIdxAuth.TokenType,
+                      completion: OktaIdxAuth.ResponseResult<OktaIdxAuth.Response>?)
+}
+
+protocol OktaIdxAuthImplementationDelegate: class {
+    func didSucceed(with token: IDXClient.Token)
+    func didFail(with error: Error)
+}
+
+protocol OktaIdxAuthRemediationRequest {
+    func send(to implementation: OktaIdxAuth.Implementation,
+              from response: IDXClient.Response)
+}
+
+extension OktaIdxAuth {
+    class Implementation {
+        let client: IDXClient
+        var delegate: OktaIdxAuthImplementationDelegate?
```

Comment:
Should be `weak`?

Reply:
It will be eventually. Right now it's more convenient this way, since without a strong reference the client gets deallocated the moment it goes out of scope. The wording should possibly change.